### ISSUE TITLE
Nvidia GPU should be supported in K8s, and the feature could be added like this.

### ISSUE THREAD

**Issue by Hui-Zhi** [kubernetes/kubernetes#19049]
Please let know if this design is OK for GPU in K8s or you need more details.
# 
# Requirements:

GPU is important to lots of workloads, such as machine learning, video stream decoding/coding, etc... But Kubernetes doesn't support schedule/tracking it which need to be.
# 
# Design:

To support GPU in K8s, there's 2 capabilites need to be enabled,

```
1. Monitoring.
    1). Kubelet need to provide the GPU hardware information along with the CPU, memory information to the Master. 
    2). Each GPU only have 2 status, in using or not in using. This is because so far GPU only exist in /sys/fs/cgroups/devices, 
        so 1 GPU is the smallest unit we can assign to container.

2. Scheduler support.
    1). To schedule Pods based on GPU, scheduler need to know the below information:
        a. How many GPUs and what types are they on each Node.
        b. What the status (in/not in using) of the GPUs on each Node.
    2). Device mapping. Once know which Node have GPUs could be used, need to map the right GPUs to container, like the "--device" in Docker. To mapping the GPU to
        container, we may need to know below things.
        a. the phsical devices, like /dev/nv0, /dev/nv1 ... This information should be provided by kubelet, kubelet on each Node provides GPU information. And I think 
           we also need to provide more information to ensure the app in container could works fine with the assgined GPU, information like "Vendor ID/Device ID, 
           subsystem_device/sybsystem_vendor".
```
# 
# YAML example:

apiVersion: v1
kind: Pod
metadata:
  name: my-nginx
spec:
  containers:
- name: nginx
  image: nginx
  ports:
  - containerPort: 80
    devices:
      name: gpu
      type: nvidia
      number: 1
    devices:
    name: gpu
    type: nvidia
    number: 2

**Comment by therc** [comment]
Reserving one GPU to one pod is a start, but it should also be possible to pack several jobs on the same GPU, if the sum of their video memory requirements fit what the GPU offers, to increase utilization. There are limits in terms of cores each GPU provides, of course, but there also seem to be limits in the runtime as well. This paper claims that as of a few years ago you couldn't share a GPU more than eight ways: http://web.missouri.edu/~becchim/tenure_documents/becchi_hpdc2012.pdf

Also see the more recent http://arxiv.org/pdf/1511.07658.pdf on GPU sharing.

We currently schedule CUDA jobs in a very crude fashion on bare hardware, using a whole GPU per job; we're hoping that with Kubernetes we could improve efficiency by increasing concurrency.

If there's a concern over security, it might be OK at first to forbid different users from getting scheduled on the same GPU. (I'm no expert on their security)

**Comment by erictune** [comment]
@kubernetes/goog-control-plane @erictune

**Comment by Hui-Zhi** [comment]
@therc , I think we can enable the GPU as a normal device which under /sys/fs/cgroup/devices first, then we have 2 choices,
1. Enable other devices which also under sys/fs/cgroup/devices.
2. Enable GPU in cgroup (kernel space), then we can enable whatever functionalities we want.

**Comment by davidopp** [comment]
@cameronbrunner

**Comment by therc** [comment]
The document also doesn't seem to address the fact that, in addition to /dev/nvidiaX, nVidia boards also need /dev/nvidiactl and /dev/nvidia-uvm.

**Comment by vishh** [comment]
At a meta level, device information need not be exposed outside of the kubelet. Once a pod that needs a GPU lands on a node, kubelet can do everything necessary to provide access.

Now for sharing GPUs, why not run multiple containers per pod for now?

**Comment by therc** [comment]
Sharing GPUs through multiple containers will work for our API use case, where we have long-running servers behind the endpoints, but is not really going to work for batch pipelines that perform model training. Some of them run for minutes, some for days.

**Comment by Hui-Zhi** [comment]
@therc Yes, for Nvidia GPU, needs "--device /dev/nvidia0:/dev/nvidia0 --device /dev/nvidiactl:/dev/nvidiactl --device /dev/nvidia-uvm:/dev/nvidia-uvm", and so far I don't know what the special works to do for other GPUs. We need to list the vendor in the yaml file.

**Comment by erictune** [comment]
Three design alternatives to consider: Renumber Devices, selectively expose devices, or expose all devices.

- Renumbering
  - when application is allocated `/dev/nvidia2`, it sees `/dev/nvidia0` because kubelet does `docker run --device /dev/nvidia2:/dev/nvidia0 ...`.
   - does this even work, or do applications need to know their number to use `/dev/nvdiactl` and `/dev/nvidia-nvm`?
   - might help with running applications that are hardcoded to use all available GPUs?
- Selectively Exposing
  - when application is allocated `/dev/nvidia2`, it sees only `/dev/nvidia2`.  It cannot see devices it is not allocated.  Like `docker run --device /dev/nvidia2:/dev/nvidia2`.
  - all pods that use a `/dev/nvidia*` device need access to  `/dev/nvidiactl` and `/dev/nvidia-nvm` too.  
  - is there any benefit to hiding some devices, given that pods can probably hurt each other via the shared devices ( `/dev/nvidiactl` and `/dev/nvidia-nvm`)?
- Expose All
  - When application is allocated a GPU, it is told which GPU, via e.g. an env var.  It can see all GPUs, but is expected to only use its own ones.
  - simpler to implement.
  - could just use hostDir to expose `/dev` to pods that need GPUs instead of special-casing  `/dev/nvidiactl` and /dev/nvidia-nvm` in kubelet.

**Comment by erictune** [comment]
Another set of alternatives, this time for reporting node capacity:
- Kubelet detects: 
  - kubelet detects presence of devices like `/dev/nvidia*` and post number and/or type present to the apiserver (e.g. in the Node object).
  - least work to setup, but hardest to customize.
- Kubelet proxies:
  - node admin is responsible for writing the device type/count into a file, which kubelet reads and reports to apiserver.
- Custom reporter:
  - cluster admin runs a pod on every node (e.g. using DaemonSet) which detects and reports the number and/or type of devices present to the apiserver
  - most flexible.
  - both kubelet and custom reporter need to avoid clobbering each other's updates if they are updating the same object -- need to agree on which field belong to which client.
  - custom reporter pods are root equivalent (maybe this is okay)

**Comment by erictune** [comment]
I think it would be a good test of the extensibility of Kubernetes to see if this feature can be implemented (initially) without any GPU-specific changes to Kubernetes.

I think this is almost possible.  If we use the "Expose All" approach I mentioned above, then kubelet does not need to change how it creates pods.

If we use the "Custom Reporter" approach I mentioned above, then the kubelet does not need to change how it reports resource capacity to the scheduler.

The [Scheduler Extender](https://github.com/kubernetes/kubernetes/blob/master/docs/design/scheduler_extender.md) feature could be used to filter nodes whose capacity is less than the requirements of the already-scheduled pods plus the candidate pod.

The main things that is missing is a way for a scheduler extender to modify a pod to allocate a specific GPU number to it.  


@ravigadde @davidopp Can a scheduler extender design be expanded to allow the scheduler extender to put an annotation on a pod?  That might allow it to be used to implement GPU scheduling.

**Comment by ravigadde** [comment]
@erictune

The extender can return annotations to the scheduler. The scheduler can update the annotations in the binding object to propagate these back to the pod. However, this will require changes in the scheduler interface.

```go
type ScheduleAlgorithm interface {
	Schedule(*api.Pod, NodeLister, SchedulerExtender) (selectedMachine string, err error)
}
```

Along with selectedMachine, it should be possible to return annotations.

**Comment by davidopp** [comment]
We really want to limit scheduler extender to Ravi's use case. I will read this issue more thoroughly in the next day or so but we don't want to turn scheduler extender into something required to support a general feature like GPUs. If that's not what you were implying, my apologies, I will read the issue more carefully soon.

**Comment by Hui-Zhi** [comment]
So far GPU only exist in /sys/fs/cgroup/devices, so any user cases should follow the rules from:
https://www.kernel.org/doc/Documentation/cgroups/devices.txt
Due to constrained by this rule, GPU only can be used by exclusive/sharing among containers. So there should have a reference counter for each GPU resource, 0 for "reference" means no containers in using, otherwise, means there're "reference" number containers are using this GPU in parallel.

Each GPU should be able to bind to more than 1 containers, but there should have a threshold value for the max number of each series GPU could sustain containers.


I just configured a server with 2 NVIDIA cards, and will do some tests.
lspci | grep NVIDIA
05:00.0 VGA compatible controller: NVIDIA Corporation GM107GL [Quadro K2200] (rev a2)
05:00.1 Audio device: NVIDIA Corporation Device 0fbc (rev a1)
0b:00.0 VGA compatible controller: NVIDIA Corporation GM107GL [Quadro K2200] (rev a2)
0b:00.1 Audio device: NVIDIA Corporation Device 0fbc (rev a1)

**Comment by Hui-Zhi** [comment]
@davidopp Could you also assign me as owner or assignee?

**Comment by davidopp** [comment]
OK, I read through the issue.

I like the idea of seeing if we can handle this in a generic fashion, but I would suggest a somewhat different approach than what's suggested above. I think we should make this a trial of the "generic/opaque counted resource" idea we've talked about but never implemented. So Kubelet would have to know something about GPUs, but we'd structure that configuration information in a way such that the same mechanism can be reused when we add other types of counted resources in the future.

One way to do this would be to have some kind of config file on kubelet (ultimately a ConfigResource (#18455) once all config is handled as API objects) with some information about each of these generic/opaque resources. What information would we need? For the GPU case IIUC we would need
* name (presumably "gpu")
* a mechanism (e.g. a hard-coded number or a shell command to execute) to determine the capacity of the resource
* a mechanism (e.g. a shell command to execute) to determine the number of units of the resource that are free (or, conversely, in use)--maybe not strictly necessary but useful for metrics API
* a command to execute when a container in a pod that requests the resource is started up on kubelet; parameterized by index(es) of the counted resource that the container should use (this could accommodate either passing the index(es) as an env var, as in Eric's "expose all," or on the docker command line, as in Eric's "selectively exposing"). The index(es) would be computed simply as "if the pod requests N units of the resource, and M are already allocated, then M, M+1, ..., M+N-1" -- hopefully we wouldn't need to support tricky allocation schemes. (We wouldn't try to reuse this mechanism for NUMA node allocation or anything like that.)

The scheduler would only have to know about the concept of "generic/opaque counted resources," not the details of any resource in particular. It would schedule against capacity of the resource as reported in NodeStatus.Capacity (soon to be called NodeStatus.Allocatable).  

Does this sound reasonable?

(This is related to #19082 which is for non-consumable resources. @cameronbrunner)

**Comment by therc** [comment]
I'm trying to get some answers about the low level details from nVidia folks, but I just remembered yet another complication: the kernel driver has one ABI it supports and the runtime is pretty strict about it matching. Which means that if you need to switch an image from CUDA 6 to CUDA 7, now you need both kinds of machines in the cluster. Unless you rebuild all images at once, bring down v6 jobs and start v7 ones. That's fairly draconian.

Since we're talking about kernel modules, you CAN unload and reload different versions of the same driver, thus avoiding a reboot. I wouldn't expect Kubernetes/Kubelet to do this kind of juggling, of course, but it could provide a few hooks.

I can see two ways to do this, keeping with the example of v6 and v7:

 1. make a given node serve only a specific version. So, when the first user shows up with the need for CUDA 7, their pods will stay pending. The user contacts the administrator, who will pick a machine or two, drain any CUDA 6 jobs on them and convert them to CUDA 7 systems (using some tool that unloads the old modules and loads the new ones). Whatever reporting runs on the node will pick up the new version. Given historical workloads and production needs, users and administrators work out a schedule to move everyone to CUDA 7, converting batches of machines to reduce disruption (you could call this "steamrolling"...). Or perhaps everyone is happy if the cluster is 60% on v6 and 40% on v7, etc.

2. make nodes report capacity for multiple versions, but as soon as a pod running v6 is scheduled, the capacity for v7 disappears (and viceversa). When the pod is done, both versions are reported again. Bonus points for scheduling pods on fewer nodes so that there are free machines which can serve either. The administrator would still have control on how many nodes have v6, v7 or both.

Other than that, @davidopp's approach sounds reasonable.

**Comment by davidopp** [comment]
@therc I agree we don't really want Kubernetes to be responsible for installing the right kernel driver (though it is a feature one could imagine supporting in the future). Option (a) sounds reasonable -- the counted-resource name would be different depending on which ABI version is supported on that machine. This suggests that maybe the first two bullets in my list should be combined.

It's possible that (b) might be needed for other features in the future (any time there is more than one way to request the same underlying resource), but I think it's better not to introduce that complication until we find something that absolutely requires it.

**Comment by Hui-Zhi** [comment]
@davidopp , Regarding your 

a mechanism (e.g. a hard-coded number or a shell command to execute) to determine the capacity of the resource
a mechanism (e.g. a shell command to execute) to determine the number of units of the resource that are free (or, conversely, in use)--maybe not strictly necessary but useful for metrics API

Could we just combine them, and just give a weight for all the same devices on a Node?
Like we have 1 Node with 2 GPUs, and assume the weight is 200, Pod A may consume 20 weight GPU, Pod B need 50 weight GPU. So may need to configure the usage in each Pod's yaml. The allocation will be handled in kubelet, and kubelet only let scheduler know there's still 170 weight of this GPU is free by NodeStatus.Allocatable.

**Comment by davidopp** [comment]
I don't think it works for scheduler to use free reported by Kubelet because scheduler would have to wait for update from kubelet in order to see resource consumption of pods it scheduled [so if it scheduled a pod that fills the whole machine, it might schedule another pod that fills the whole machine right afterwards, before hearing back from kubelet]. We have to store requested resources in etcd, anyway so I don't see any reason why it would be better to use report from Kubelet instead of request that is in the pod description that was written to etcd.

In the future we might want to do something like Borg's "resource estimation" where we schedule based on some value the master synthesizes from request and usage, which is closer to what you are suggesting.

On an unrelated, note I realized another reason (besides monitoring) why we need a way to determine the number of the units that are free/in use on kubelet is so kubelet can do feasibility check on admission.

**Comment by Hui-Zhi** [comment]
I have run some NVIDIA CUDA testings, they shows 1 application in container at least need to be assigned to 1 GPU, 50% doesn't work, and multiple containers can be assigned to multiple GPUs, they can be overlapped. Also I have took a look at NVIDIA-Docker, it shows the CUDA version inside container must equal the CUDA on the host, they use docker image label to ensure that.

Regarding the implementation in both scheduler and kubelet, I think it could be like this.
scheduler:
the data struct could be defined as:
type HWRequirement struct {
  	Name string  //the GPU name, also should included the vendor info. like NVIDIA-CUDA7.5-GPU.
	Count int    // how many the same name GPU have on the node.
}

kubelet:
kubelet should be the one know all the GPU details. Like GPU paths (/dev/nvidia0, /dev/nvidia1), venders, even the framework version perhaps, also need to be able to assigned these devices to container, and each kind of hardware should have  its own way to bind itself to container.

Regarding the kubectl:
it should write the GPU requirement also like vender-frameworkversion-GPU.

**Comment by erictune** [comment]
@davidopp : does kubelet have to run a shell script to determine the number of units of the resource that are free?  Or can it do resource math across all the containers it sees scheduled to it?

@Hui-Zhi As I understand it, GPUs (or at least most) do not support preemptive multitasking.  So, I don't see the benefit of the weight approach.

**Comment by davidopp** [comment]
> does kubelet have to run a shell script to determine the number of units of the resource that are free? Or can it do resource math across all the containers it sees scheduled to it?

I don't see how to avoid something that's fully general like a shell script. You can't get usage out of the container description; you need to pull it out of the kernel somehow. For reporting capacity, you always need something fully general. At least I do not have a better idea.

**Comment by erictune** [comment]
We only need to report usage if we want to overcommit a resource.  If the resource can only be used in a guaranteed fashion, then I guess a script is not needed?  I think gpu and all the other use cases i can think of fit the guaranteed only model.

**Comment by therc** [comment]
https://github.com/NVIDIA/nvidia-docker/wiki has two ways of running containers, using either a wrapper around the docker command or by running a Docker plugin (which exposes both devices and volumes).

I'm assuming the latter is more relevant to us. It exposes [a lot of device specs and status information, including affinity](https://github.com/NVIDIA/nvidia-docker/blob/master/tools/src/nvidia-docker-plugin/remote_v1.go), then can return the Docker command line flags to map the nvidiactl, nvidia-uvm and nvidiaX..Y devices, as well as volumes. It has Mesos support, too! The volume part of it is also careful about [which binaries and which libraries to export](https://github.com/NVIDIA/nvidia-docker/blob/master/tools/src/nvidia/volumes.go). It looks to me like it offers what's necessary for both Phase 1 (guaranteed resources) and Phase 2 (overcommitting).

**Comment by Hui-Zhi** [comment]
nvidia-smi tool can get all the information we want, but if you run nvidia-smi inside container and container only be assigned 1 GPU, the tool could only show 1 GPU's information.

"the weight" I think may could let scheduler knows how many containers we could assign to 1 GPU.

**Comment by Hui-Zhi** [comment]
I think to implement GPU device in kubelet, we should implement the GPU as kubelet plugins, like:
allPlugins = append(allPlugins, nvidia.ProbeVolumePlugins()...)
allPlugins = append(allPlugins, amd.ProbeVolumePlugins()...)
allPlugins = append(allPlugins, intel.ProbeVolumePlugins()...)

So far I'm working on the NVIDIA's GPU, and in kubelet could get the below information,
gpuDevices:{GPUPlatform:{Name:cuda DriverVersion:352.63 PlatformVersion:7.5} Devices:[{Path:/dev/nvidia0 Cores:640 Memory:4092 Family:Maxwell}]}

Actually, by my understanding, we can talk about the features and design by opening a new issue, but regarding the detail design, should I write a spec like openstack does to let you  review? Or I write the code in my fork, and let you review the code directly? I have read the Kubernetes Developer Guide, but there's no details about this, also asked in the kubernetes-dev slack, also got no answers.
@davidopp  @erictune

**Comment by erictune** [comment]
Writing a design doc first is a good idea.  Write a document called `docs/design/gpu.md` and send a PR to add that doc.  People can comment on the PR of the doc, before you write any code.

**Comment by Hui-Zhi** [comment]
Sorry for the late reply, just came back from Chinese Spring Festival, have write the rough design doc at https://github.com/Hui-Zhi/kubernetes/blob/a56c9051df010390a1da7d7032fdf983eae715f3/docs/design/devices.md, would you mind let me know how to send a PR to add this doc? I can't find the email address and can't commit a change. Thanks in advance. @erictune   @davidopp

**Comment by Hui-Zhi** [comment]
Also, do you think add more details to this doc is appropriate? If so, I would like to do it.

**Comment by davidopp** [comment]
I'd like to see us do this using the "opaque integer resource" approach I described earlier in this issue ( https://github.com/kubernetes/kubernetes/issues/19049#issuecomment-170292193 ) if at all possible. If we decide that GPUs are too weird/unique to fit into the "opaque integer resource" model then I'm totally fine with doing it as a special snowflake, but I'd like us to at least explore whether it would be possible .

**Comment by davidopp** [comment]
(Oh, and I should say, I stole the idea from @erictune, who suggested the same general idea here https://github.com/kubernetes/kubernetes/issues/19049#issuecomment-168898400 )

**Comment by Hui-Zhi** [comment]
Ok, I'll change it as you suggested, do you think I need to add more details about the implement? @davidopp

**Comment by therc** [comment]
We're at the point where we could actually use the feature. We don't mind running 1.2 + patches (we've been running from HEAD for the past couple of months...) or writing a good chunk of the code.
Initially, for V1 of the support we're OK with just detecting one GPU per machine and scheduling pods on that. But we do need to overcommit, using something like [MPS (pdf)](https://docs.nvidia.com/deploy/pdf/CUDA_Multi_Process_Service_Overview.pdf). We have models that either are tiny or don't see traffic for days, other than health checks. Giving those one GPU is a huge waste.
For V2, we'd need detection of multiple GPUs. The scheduler should be smart enough to spread pods across all GPUs, not just maxing out the first device, then the second, etc. (again, we'd be overcommitting the devices. Through monitoring, we can find out if the cluster is underprovisioned or if certain models are heavier users than their nominal requirements imply.) It would work a bit like Borg disk resources: find the first device on the machine with enough resources available.

@Hui-Zhi try [comparing changes](https://github.com/kubernetes/kubernetes/compare/master...Hui-Zhi:devices_support) to get a 'Create pull request' button.

**Comment by therc** [comment]
I don't know that the overcommitting matters to most people out there, so I'll create a separate proposal for a scheduler extender that does that.

**Comment by Hui-Zhi** [comment]
@therc Thanks let me know that, I'll take a look at the pdf you provided first.

@davidopp I agree with do this using the "opaque integer resource", but it's not my idea, so I just commit the original file I wrote, I think after you add it as comment to this doc, then I commit it again is better. Is this OK?

**Comment by Hui-Zhi** [comment]
I have done some tech testing and verification. The code at https://github.com/Hui-Zhi/kubernetes/tree/try_nvidia_docker_merge_devices_support already could schedule GPU resources, but it's really really ugly!!! It refer some codes from NVIDIA-Docker. I'll reconstruct them and then show you the way I implement it. @erictune , Could you help review the design which at:
https://github.com/kubernetes/kubernetes/pull/21446

**Comment by Hui-Zhi** [comment]
I think if we add GPG features in K8s, the binaries compiled by K8s would need GPU dependences, like NVIDIA GPU GDK, is this ok?

**Comment by therc** [comment]
I think there might be some resistance to linking in new code, especially if it's large (is it? I haven't checked) and very few users will take advantage of it. In my v0 proposal, we use the dependencies in additional binaries.

**Comment by davidopp** [comment]
https://github.com/NVIDIA/nvidia-docker/

**Comment by therc** [comment]
BTW, I opened an issue there and they seem receptive to either adding new endpoints or letting us link the `nvidia` package directly: https://github.com/NVIDIA/nvidia-docker/issues/66

**Comment by therc** [comment]
@Hui-Zhi: you already experimented with a lot of the work in your changes at https://github.com/kubernetes/kubernetes/compare/master...Hui-Zhi:try_nvidia_docker_merge_devices_support

Initially we might have to

- keep nvidia-docker-plugin in an external process (which also means no more giant Godeps changes) and define additional HTTP endpoints for it
- add the resource `nvidiaGpu` instead of `devices`
- split the whole set of changes into a number of smaller, separate PRs

How do you want to proceed?
Once we find the code to be stable enough and perhaps after it has been made more generic, we can work to build into the kubelet, which will look a lot like the code you have now.

**Comment by Hui-Zhi** [comment]
@therc I think just add "--device" to kubelet, and all the GPU allocate/release things let Kubelet do is the best way. Regarding add the resource nvidiaGPU instead of devices, it's OK for now, but it would be better we add devices as the name, and in the devices, we can add GPU, then the driver of GPU also need to be described, like "nvidiaGPU", "IntelGPU" ... and the need to change the current struct of resources. Currently regarding CPU/Memory, only have number. But regarding other resources, it could have hierarchies.

**Comment by Hui-Zhi** [comment]
By the way, I'll update my PR and include the answers of @erictune 's questions about it.

**Comment by Hui-Zhi** [comment]
Have update the code of scheduler which could support GPU as a first-class integer resource. The code has been tested, and will submit the kubelet part later.

@therc @davidopp @vishh @erictune @hongchaodeng @timothysc 

https://github.com/kubernetes/kubernetes/pull/21446/files

**Comment by hsaputra** [comment]
Should this issue be renamed to indicate it will be primarily support for NVDIA GPUs?

**Comment by therc** [comment]
@hsaputra are you one of those mythical non-NVIDIA GPU users?

Jokes aside, we have been wondering about other use cases. Can you tell us more about what you'd use  GPUs for, under Kubernetes?

**Comment by hsaputra** [comment]
LOL, no we also use NVIDIA. I just wanted to make sure that the issue reflect the right problem and potential solution to solve it.

Will be using it to run ML type of programs in containers managed by (hopefully) Kubernetes.

**Comment by Hui-Zhi** [comment]
@hsaputra so far it only about the Nvidia GPU. So, yes.

**Comment by hsaputra** [comment]
Thanks @Hui-Zhi

**Comment by dirtGit** [comment]
@Hui-Zhi "gpuDevices:{GPUPlatform:{Name:cuda DriverVersion:352.63 PlatformVersion:7.5} Devices:[{Path:/dev/nvidia0 Cores:640 Memory:4092 Family:Maxwell}]}" How to get these GPU info?

**Comment by georghildebrand** [comment]
Hey all, i am reading this with a lot of interest. From the perspective of a scientific user of GPUs i can also recommend to consider adding socket / cpu affinity information from the kubelet  to not slow down performance for shared workloads. Especially on gpu grid tasks this will be helpfull and speeds up the adoption of kubernetes in scientific environments. 

Example:
`$nvidia-smi topo -m
	GPU0	GPU1	GPU2	GPU3	CPU Affinity
GPU0	 X 	PIX	SOC	SOC	0-7,16-23
GPU1	PIX	 X 	SOC	SOC	0-7,16-23
GPU2	SOC	SOC	 X 	PIX	8-15,24-31
GPU3	SOC	SOC	PIX	 X 	8-15,24-31

Legend:

  X   = Self
  SOC  = Connection traversing PCIe as well as the SMP link between CPU sockets(e.g. QPI)
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe switches (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing a single PCIe switch
  NV#  = Connection traversing a bonded set of # NVLin`

**Comment by timothysc** [comment]
@georghildebrand  we're working on it.. /cc @jeremyeder
